src/libraries/System.Private.CoreLib/src/System/Collections/Generic/ReferenceEqualityComparer.cs

```diff
@@ -0,0 +1,64 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.CompilerServices;
+
+namespace System.Collections.Generic
+{
+    /// <summary>
+    /// An <see cref="IEqualityComparer{Object}"/> that uses reference equality (<see cref="object.ReferenceEquals(object?, object?)"/>)
+    /// instead of value equality (<see cref="object.Equals(object?)"/>) when comparing two object instances.
+    /// </summary>
+    /// <remarks>
+    /// The <see cref="ReferenceEqualityComparer"/> type cannot be instantiated. Instead, use the <see cref="Instance"/> property
+    /// to access the singleton instance of this type.
+    /// </remarks>
+#if INTERNAL_REFERENCE_EQUALITY_COMPARER
+    internal
+#else
+    public
+#endif
+    sealed class ReferenceEqualityComparer : IEqualityComparer<object?>, IEqualityComparer
```

Comment:
Move this file to Common since its shared?

Reply:
I considered that, but then I looked at how `NullableAttributes.cs` was brought in, including how they've put the `#ifdef` into the code and in which directory they've put their files. And I based this code off of the pattern established by that.

https://github.com/dotnet/runtime/blob/f99b4fe46497185db3b2049913f2af0c839fda00/src/libraries/Directory.Build.targets#L191-L201

This also makes more sense to me because the "natural" place for this logic is in CoreLib, since that's where it's publicly exposed, and anybody who wants to pull it in has to take the _additional_ step of declaring the proper compile-time constants. If everything were entirely internal then I'd agree it should be in the Common folder.

Reply:
Good to know. Makes sense to me. Thanks for sharing.

Reply:
I agree that it does not make sense to move public netcoreapp types under Common to support downlevel builds.

It would be nice to use consistent pattern for this. There is variety of styles:
- `INTERNAL_NULLABLE_ATTRIBUTES` mentioned above.
- `REGISTRY_ASSEMBLY` https://github.com/dotnet/runtime/blob/4f9ae42d861fcb4be2fcd5d3d55d5f227d30e723/src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeRegistryHandle.cs#L14
- `MAKE_ABW_PUBLIC` https://github.com/dotnet/runtime/blob/4f9ae42d861fcb4be2fcd5d3d55d5f227d30e723/src/libraries/Common/src/System/Buffers/ArrayBufferWriter.cs#L12

Is the `INTERNAL_XXX` style we like the best?

Reply:
> Is the INTERNAL_XXX style we like the best?

I actually prefer the `MAKE_ABW_PUBLIC` style, albeit with a better name. My concern with `INTERNAL_XXX` is it defaults to the thing being public, so if you neglect to specify the compilation constant, you end up exposing the type by accident.  Our tooling in theory helps to avoid that, but it still makes me a little nervous.

Reply:
The reason I was hesitant to use the `MAKE_XYZ_PUBLIC` style is that if this code exists in the `CoreLib` folder, then the "natural" (default) state of the code without any extra compilation parameters being set should be such that it's appropriate for the `System.Private.CoreLib` project. I didn't want to have to make any changes to `System.Private.CoreLib.csproj` just to do the right thing.

Honestly at this point I'd rather revert all the ifdef changes in the `CoreLib` ReferenceEqualityComparer.cs file, make that type public with all of the appropriate nullability annotations, and put a copy of the file in the `Common` folder with the type marked _internal_ and the file marked `#nullable disable`. That `Common` copy would be for JSON and BinaryFormatter tests usage. Yes, it's code duplication, but it's such a trivial amount of duplication, and it keeps the file containing the production type clean.

Reply:
> if this code exists in the CoreLib folder, then the "natural" (default) state of the code without any extra compilation parameters

Maybe have `SYSTEM_PRIVATE_CORELIB` defined when building CoreLib, and then all the ones that need to be public in CoreLib and nowhere else can use this?

Here is another one that can use this: https://github.com/dotnet/runtime/blob/master/src/libraries/Common/src/System/Diagnostics/CodeAnalysis/ExcludeFromCodeCoverageAttribute.cs#L8